### DISCUSSION TITLE
how can i save my log, when bpftrace being kill -9

### DISCUSSION THREAD

**Discussion by zxc199108** [bpftrace/bpftrace#2143] [Q&A] [answered]
I run this cmd in my server to catch slab information, but some error occurred.
my bpftrace process be killed, save nothing information.
I know bpftrace can handler kill -2 , and output result.

so how can i save my result, when system kill -9 my process?

`nohup bpftrace kmalloc.bt -o testlog.txt &`

**Comment by fbs** [comment, marked as answer]
Afaik internal buffers are flushed when a 'std::endl' is written, which we use. However we do not flush mid the printing of complex values, like histograms.

Do you have an example of when this happens?

**Reply by zxc199108** [reply]
#include <linux/sched.h>
k:kmalloc_slab /arg0 == 65536/
{
        @detail[comm,arg0,kstack] = count();
}

I run this script to catch slab information, when my server have 2G free ram space.
but my server happend oom and  kill process, my terminal can not work, because no more ram.
so i have to reset my server, my bpftrace log dont save.

**Reply by fbs** [reply]
Yes that structure is not printed until bpftrace exits, which it doesn't on segfault. Terminating gracefully on SEGV seems like a tricky thing to do.

You could try adding a `interval:s:1 { print(@detail);; }` but not missing any event in a SEGV case will be difficult.

If you really really need as much info as possible I think writing a C version of this will be the way to go. That way you can focus on perisisting data to disk asap. Might even be possible to pin the maps and prog to have it persist longer than the program that created them

**Reply by zxc199108** [reply]
Maybe i have to write a C version and use bcc to compile.
Thanks fbs.
